modules/dnn/include/opencv2/dnn/dnn.hpp

```diff
@@ -1079,28 +1091,31 @@ CV__DNN_INLINE_NS_BEGIN
     /** @brief Creates 4-dimensional blob from image. Optionally resizes and crops @p image from center,
      *  subtract @p mean values, scales values by @p scalefactor, swap Blue and Red channels.
      *  @param image input image (with 1-, 3- or 4-channels).
+     *  @param scalefactor scalefactor multiplier for @p image values. It can be `double` and `Scalar`. The `Scalar`
+     *  will be multiplied from channels. And the double means it will be multiplied on all channels.
      *  @param size spatial size for output image
      *  @param mean scalar with mean values which are subtracted from channels. Values are intended
      *  to be in (mean-R, mean-G, mean-B) order if @p image has BGR ordering and @p swapRB is true.
-     *  @param scalefactor multiplier for @p image values.
      *  @param swapRB flag which indicates that swap first and last channels
      *  in 3-channel image is necessary.
      *  @param crop flag which indicates whether image will be cropped after resize or not
-     *  @param ddepth Depth of output blob. Choose CV_32F or CV_8U.
+     *  @param ddepth Depth of output blob. Choose CV_32F, CV_16F or CV_8U.
      *  @details if @p crop is true, input image is resized so one side after resize is equal to corresponding
      *  dimension in @p size and another one is equal or larger. Then, crop from the center is performed.
      *  If @p crop is false, direct resize without cropping and preserving aspect ratio is performed.
      *  @returns 4-dimensional Mat with NCHW dimensions order.
+     *
+     *  @note The calculation order of @p mean and @p scalefactor is `(image - mean) * scalefactor`.
      */
-    CV_EXPORTS_W Mat blobFromImage(InputArray image, double scalefactor=1.0, const Size& size = Size(),
+    CV_EXPORTS_W Mat blobFromImage(InputArray image, const Scalar& scalefactor = 1.0, const Size& size = Size(),
```

Comment:
Currently, this patch breaks the compatibility of the java bindings.

Reply:
It is expected.
As this patch modifies functions instead of adding overloads.

Reply:
> const Scalar& scalefactor = 1.0

Default ("implicit" ctor) value may not work as expected.

There are issues of Scalar initialization:

```.cpp
Scalar a(2.0);
cout << a << endl;
Scalar b = Scalar::all(3.0);
cout << b << endl;
```

```
[2, 0, 0, 0]
[3, 3, 3, 3]
```

Reply:
Yes, I addded code in this patch to convert from `Scalar a(2.0);` to `Scalar a = Scalar::all(2.0);`:
```c++
Scalar scalefactor = param.scalefactor;
//  Because 0 is meaningless in scalefactor. Convert the 0 to
if (scalefactor[1] == 0 && scalefactor[2] == 0 && scalefactor[3] == 0)
{
    CV_Assert(scalefactor[0] != 0 && "Scalefactor of 0 is meaningless.");
    scalefactor = Scalar::all(scalefactor[0]);
}

```